### ISSUE TITLE
Using TabBar within CupertinoTabScaffold

### ISSUE THREAD

**Issue by adriancmurray** [flutter/flutter#18275]
I would like to use the properties of a TabBar (scrolling between views, an indicator, and material design) but within a CupertinoTabScaffold. I want this so I can run parallel navigation hierarchies while also rendering each TabView fully like what you'd expect in iOS. I know this isn't going to be conducive to Material Design. I know some people have done this with taking out _TabSwitchingView but I'm struggling to make it happen. @xster  @HansMuller  or @Hixie ? I see you three are pretty much the team on here to ask... Think you could help direct me a little? Thank you.

**Comment by Hixie** [comment]
@xster we should make sure that these widgets are self-contained enough that it's trivial to swap one out for the other, if we haven't already done that.

**Comment by xster** [comment]
The power of parallel navigation hierarchies actually doesn't come from CupertinoTabScaffold. CupertinoTabScaffold is quite simply a Stack that places a tab bar at the bottom which you can easily recreate yourself or use standard Material Scaffold in its place.

Parallel navigation rather comes from the usage of CupertinoTabView inside each tab (you can refer to the [CupertinoTabScaffold doc's sample code](https://docs.flutter.io/flutter/cupertino/CupertinoTabScaffold-class.html)). 

Making those the content of each Material Scaffold with Material TabBar's tab will achieve the effects you want.

**Comment by xster** [comment]
Besides the parallel navigation, we can also extract _TabSwitchingView out of Cupertino into the Widgets layer for some extra reusable tab switching utility. But it's very basic since it just flips between one tab and another within one frame and is fairly iOS design specific. 

See the Material BottomNavigationDemo for a simple alternative implementation with animation https://github.com/flutter/flutter/blob/d803f02d4f59bf06f4d3c26a6fa9669843908405/examples/flutter_gallery/lib/demo/material/bottom_navigation_demo.dart#L172

**Comment by adriancmurray** [comment]
@xster THANK YOU! I was actually looking into CupertinoTabView last night but hadn't implemented it correctly. Your phrasing helped me put things in the right place. Thank you so much for that! Now that you helped me with that I ran into the same issue as #17555. I'll wait on fixing that problem and may not even use a BAB or FAB, it'd just be nice to. I do have one last question regarding the TabBar and implementing the CupertinoTabView for each page: is there a way to force the state of each tab to remain? When I leave from one tab to the next the page is redrawn and the navigation history lost. I'm sure this is a limitation of TabBar since it's probably not designed to act like CupertinoTabScaffold. Thank you again!

**Comment by xster** [comment]
It's not so much a limitation of the TabBar since the TabBar just graphically represents the bottom of the screen. 

You need to keep the ancestors of the CupertinoTabView in the tree via Offstage on a Stack or something similar. If the CupertinoTabViews are removed, then ya, the Navigator states would be lost. 

This is what the _TabSwitchingView does. If you'd like to send a PR to make _TabSwitchingView public, I think it could be sensible.

**Comment by adriancmurray** [comment]
Thank you @xster I've done that (though without a description) here: #18335

I'll close this now.
